pkgs/development/compilers/flutter/update/update.py

```diff
@@ -0,0 +1,340 @@
+        if not line:
+            break
+        stderr += line
+        print(line.strip())
+
+    process.wait()
+    temp.close()
+    return stderr
+
+
+def get_artifact_hashes(flutter_compact_version):
+    code = load_code("get-artifact-hashes.nix",
+                     nixpkgs_root=NIXPKGS_ROOT,
+                     flutter_compact_version=flutter_compact_version)
+
+    stderr = nix_build_to_fail(code)
+
+    pattern = re.compile(
+        r"/nix/store/.*-flutter-artifacts-(.+?)-(.+?).drv':\n\s+specified: .*\n\s+got:\s+(.+?)\n")
+    matches = pattern.findall(stderr)
+    result_dict = {}
+
+    for match in matches:
+        flutter_platform, architecture, got = match
+        result_dict.setdefault(flutter_platform, {})[architecture] = got
+
+    def sort_dict_recursive(d):
+        return {
+            k: sort_dict_recursive(v) if isinstance(
+                v, dict) else v for k, v in sorted(
+                d.items())}
+    result_dict = sort_dict_recursive(result_dict)
+
+    return result_dict
+
+
+def get_dart_hashes(dart_version):
+    platforms = [
+        "x86_64-linux",
+        "aarch64-linux",
+        "x86_64-darwin",
+        "aarch64-darwin"]
+    result_dict = {}
+    for platform in platforms:
+        code = load_code(
+            "get-dart-hashes.nix",
+            dart_version=dart_version,
+            platform=platform)
+        stderr = nix_build_to_fail(code)
+
+        pattern = re.compile(r"got:\s+(.+?)\n")
+        result_dict[platform] = pattern.findall(stderr)[0]
+
+    return result_dict
+
+
+def get_flutter_hash_and_src(flutter_version):
+    code = load_code(
+        "get-flutter.nix",
+        flutter_version=flutter_version,
+        hash="")
+
+    stderr = nix_build_to_fail(code)
+    pattern = re.compile(r"got:\s+(.+?)\n")
+    hash = pattern.findall(stderr)[0]
+
+    code = load_code(
+        "get-flutter.nix",
+        flutter_version=flutter_version,
+        hash=hash)
+
+    return (hash, nix_build(code))
+
+
+def write_data(
+        nixpkgs_flutter_version_directory,
+        flutter_version,
+        engine_hash,
+        dart_version,
+        dart_hash,
+        flutter_hash,
+        artifact_hashes,
+        pubspec_lock):
+    with open(f"{nixpkgs_flutter_version_directory}/data.json", "w") as f:
+        f.write(json.dumps({
+            "version": flutter_version,
+            "engineVersion": engine_hash,
+            "dartVersion": dart_version,
+            "dartHash": dart_hash,
+            "flutterHash": flutter_hash,
+            "artifactHashes": artifact_hashes,
+            "pubspecLock": pubspec_lock,
+        }, indent=2).strip() + "\n")
+
+
+def get_pubspec_lock(flutter_compact_version, flutter_src):
+    code = load_code(
+        "get-pubspec-lock.nix",
+        flutter_compact_version=flutter_compact_version,
+        flutter_src=flutter_src)
+
+    stderr = nix_build_to_fail(code)
+
+    pattern = re.compile(r"For full logs, run '+(.+?)'\.\n")
```

Comment:
You need to call `nix log <derivation's .drv file>. Which I don't know of another way to get aside from getting it from the user messages.

Does this comment for a stable api also apply to the `got: <hash>` messages this script is parsing? Since I don't know of another way of getting it and it's what other update script/tools also do.

I've ended up using a derivation over calling pub locally because:
- Nothing will be written to the user's home folder in `~/.pub`
- I didn't want to have to copy and paste the entire flutter repository to somewhere local from nix store and deal with permissions

Not sure if these are good reasons, but if we can get `pubspec.lock` from the tarball I think that is the preferable way.. What do you think?

Reply:
> You need to call `nix log <derivation's .drv file>. Which I don't know of another way to get aside from getting it from the user messages.

`nix-instantiate` should do it - instead of `nix-build`, use `nix-instantiate`, grab the `.drv` path, build it with `nix-store --realise`, and check the logs with `nix log`.

> Does this comment for a stable api also apply to the got: <hash> messages this script is parsing? Since I don't know of another way of getting it and it's what other update script/tools also do.

I suppose it does, but I don't know any other way either.

> Not sure if these are good reasons, but if we can get pubspec.lock from the tarball I think that is the preferable way.. What do you think?

The tarball is more of a last resort as it has the potential to download quite a lot.

I'm thinking, when you use a blank hash and it spits out the real one, it still keeps that output in the store, right? Because when you rebuild with the right hash it's instant. So I wonder if we can make the FOD output the `pubspec.lock`. and use the hash from the error message to obtain the true output path. Then we can just copy the `pubspec.lock` from there.

Reply:
Doing in a derivation avoids the mess of filling `~/.pub` for example and forces it to download the dependencies.

BTW another, simpler, approach is the following:
- Get the derivation path using `nix eval .#package.drvPath`
- Build the derivation with `nix-store -r /nix/store/thatderivation.drv`
- Get the logs of the build using `nix log /nix/store/thatderivation.drv` then the rest is parse work basically.

Reply:
Ahh I'm not sure which way is better.
For now I'll change to the @hacker1024 's method since we already do it this way to get Flutter's source.